src/test/rpc/Simulate_test.cpp

```diff
@@ -0,0 +1,819 @@
+#include <xrpld/rpc/CTID.h>
+#include <xrpl/protocol/ErrorCodes.h>
+#include <xrpl/protocol/STBase.h>
+#include <xrpl/protocol/STParsedJSON.h>
+#include <xrpl/protocol/jss.h>
+#include <xrpl/protocol/serialize.h>
+
+#include <optional>
+#include <tuple>
+
+namespace ripple {
+
+namespace test {
+
+class Simulate_test : public beast::unit_test::suite
+{
+    void
+    checkBasicReturnValidity(
+        Json::Value& result,
+        Json::Value& tx,
+        int expectedSequence,
+        std::string expectedFee)
+    {
+        BEAST_EXPECT(result[jss::applied] == false);
+        BEAST_EXPECT(result.isMember(jss::engine_result));
+        BEAST_EXPECT(result.isMember(jss::engine_result_code));
+        BEAST_EXPECT(result.isMember(jss::engine_result_message));
+        BEAST_EXPECT(
+            result.isMember(jss::tx_json) || result.isMember(jss::tx_blob));
+
+        Json::Value tx_json;
+        if (result.isMember(jss::tx_json))
+        {
+            tx_json = result[jss::tx_json];
+        }
+        else
+        {
+            auto unHexed = strUnHex(result[jss::tx_blob].asString());
+            SerialIter sitTrans(makeSlice(*unHexed));
+            tx_json = STObject(std::ref(sitTrans), sfGeneric)
+                          .getJson(JsonOptions::none);
+        }
+        BEAST_EXPECT(tx_json[jss::TransactionType] == tx[jss::TransactionType]);
+        BEAST_EXPECT(tx_json[jss::Account] == tx[jss::Account]);
+        BEAST_EXPECT(
+            tx_json[jss::SigningPubKey] == tx.get(jss::SigningPubKey, ""));
+        BEAST_EXPECT(
+            tx_json[jss::TxnSignature] == tx.get(jss::TxnSignature, ""));
+        BEAST_EXPECT(tx_json[jss::Fee] == tx.get(jss::Fee, expectedFee));
+        BEAST_EXPECT(
+            tx_json[jss::Sequence] == tx.get(jss::Sequence, expectedSequence));
+    }
+
+    void
+    checkBasicReturnValidity(
+        Json::Value& result,
+        Json::Value& tx,
+        int expectedSequence,
+        XRPAmount expectedFee)
+    {
+        return checkBasicReturnValidity(
+            result, tx, expectedSequence, expectedFee.jsonClipped().asString());
+    }
+
+    void
+    testTx(
+        jtx::Env& env,
+        Json::Value& tx,
+        std::function<void(Json::Value, Json::Value&)> validate,
+        bool testSerialized = true)
+    {
+        Json::Value params;
+        params[jss::tx_json] = tx;
+        validate(env.rpc("json", "simulate", to_string(params)), tx);
+        params[jss::binary] = true;
+        validate(env.rpc("json", "simulate", to_string(params)), tx);
+        validate(env.rpc("simulate", to_string(tx)), tx);
+        validate(env.rpc("simulate", to_string(tx), "binary"), tx);
+        if (testSerialized)
+        {
+            // This cannot be tested in the multisign autofill scenario
+            // It is technically not a valid STObject, so the following line
+            // will crash
+            STParsedJSONObject parsed(std::string(jss::tx_json), tx);
```

Comment:
const

Reply:
Fixed in 1c0b3a6f8ca9b038035689b68971ad5e31ba30df